### ISSUE TITLE
Session Expiration

### ISSUE THREAD

**Issue by rickhumphries** [elmsln/elmsln#920]
Are the session expiration times between systems different? Is this something that's controlled in core? I think users seem to get automatically logged out of courses after a day (or something) while users in the CIS never seem to. Should these settings be synced?

**Comment by btopro** [comment]
hmm... is CIS the HUB that people get logged into? That isn't synced per-say but it is possible that Bakery rejects the session after a different amount of time then CIS because CIS is actually being logged into while bakery is not.

http://stackoverflow.com/questions/2186115/drupal-timeout-how-to-amend-timeout-length

this claims (and I've seen it before) that there's a setting that's in settings.php to control this, means it could be modified in shared_settings.php and all of them would accept this

**Comment by rickhumphries** [comment]
yeh the CIS is the master site (hub) that people log into directly now. I think having some shared settings for this would be a good idea though so the login/logout experience is always tied together. Otherwise you'll come back to the CIS (hub), clicked to access a course, get the 'access denied' message, log into courses site directly, get redirected back to CIS, before having to access the course again. It still works, just a bit fragmented

**Comment by btopro** [comment]
I believe we solved this via elmsln_bakery which is the built in bakery method of logins. It's install function might conflict w/ the work OB did since we added our core support after you guys.

```
<?php
/**
 * @file
 * Code for the ELMSLN Bakery feature.
 */

/**
 * Implements hook_page_delivery_callback_alter().
 */
function elmsln_bakery_page_delivery_callback_alter(&$callback, $set = FALSE) {
  // switch this page delivery callback so we can better act on it
  // this is an obscure method that helps reduce the chances of
  // incorrectly issuing an access denied message.  This appears
  // to potentially even be browser / server interpretation
  // driven so we need this to help clean up those nagging issues.
  // A common issue this avoids is a logged in user on a bakery
  // site with public access being given access denied when they
  // have logged into another system with bakery SSO but not this one
  if ($callback == 'drupal_deliver_html_page' || $callback == 'boost_deliver_html_page') {
    $callback = 'elmsln_bakery_deliver_html_page';
  }
}

/**
 * Callback for bakery to handle page delivery.
 */
function elmsln_bakery_deliver_html_page($page_callback_result) {
  // bail early on these things
  // test a series of cases when we should basically skip everything
  // see if we are getting access denied about to be issued
  // on a bakery slave domain and are not currently logged in.
  if ($GLOBALS['user']->uid == 0 && is_int($page_callback_result) && $page_callback_result == MENU_ACCESS_DENIED && !variable_get('bakery_is_master', 0)) {
    // do nothing, let the page go through if these cases match
    // because they are anonymous by design
    if (drupal_is_cli() || arg(0) == 'httprl_async_function_callback' || strpos(request_uri(), 'files/advagg_css/css') != 0 || strpos(request_uri(), 'files/advagg_js/js') != 0) {
    }
    else {
      // redirect to the user page which will auto login if bakery is detected
      // this also ensures a cleaner user experience if not logged in anywhere yet
      drupal_goto('user');
    }
  }
  // allow for boost to take us to the right place
  // boost has methods for helping handle this edge case and
  // is a commonly installed project
  if (module_exists('boost')) {
    boost_deliver_html_page($page_callback_result);
  }
  else {
    // Call original function which drupal will handle by default methods
    drupal_deliver_html_page($page_callback_result);
  }
}
```

This code in the main module auto logs in if the session cookie exists in the bakery master. Enabling this project outright would switch sites form pointing to CIS to pointing to People distro (which I know isn;t waht you want hence telling you not to install outright).

**Comment by rickhumphries** [comment]
Ah nice one. Yeh this is really useful anyway, thanks

**Comment by btopro** [comment]
we could abstract the references to CPR so they can be overridden by a setting in the shared_settings perhaps. Not sure what custom work OB did in this area but shouldn't be difficult to switch rails on it w/ a little review

**Comment by rickhumphries** [comment]
That would be a neat way of doing it. Always nice to stay true to core! As far as I know we haven't done anything majorly custom other than set it up, but will double-check that

**Comment by btopro** [comment]
We need to allow for a variable in settings.php to be allowed to override the source of bakery logins. We do people but its cause we have it, could be CIS or anything really so long as it's an authority (could be remote system too like d.o.)
`elmsln_bakery_install`

**Comment by btopro** [comment]
@rickhumphries this has details on how to use it included now. It should be able to be modified to talk to CIS instead of CPR pretty easily. I haven't tested it but it's pretty basic

**Comment by rickhumphries** [comment]
Ah great stuff, thanks for that :)

**Comment by btopro** [comment]
fixed in i96 which needs merged in for 0.4.x when we get the video player switched over
